### DISCUSSION TITLE
Colorization in PS1

### DISCUSSION THREAD

**Discussion by FOSSilizedDaemon** [magicant/yash#54] [Q&A]
Hello,

I have been working on porting my shell configuration from zsh to yash. As I started learning shell scripting using pdksh and only ever learned POSIX shell script this has been fairly easy, but there is one respect that I cannot for the life of me figure out properly in Yash; that is the prompts.

My prompt is nothing special.
```
[return_status]-(git repository/direnv) >>>
```
I have not gotten to implementing the `-(git repository/direnv)` part, but have implemented the return_status part. However, I was only able to implement it in a jankyish way. Normally, how I like to implement this is with the following.
```
_check_last_return_status() {
	## If the last command returned a zero exit status,
	## then save said status with pink coloring. Otherwise,
	## save said status with red coloring.
	if [ "${return_status}" -eq 0 ]; then
		return_status_prompt="\frt${return_status}\fd"
	else
                return_status_prompt="\fmt${return_status}\fd"
	fi

	## Print the value of $return_status to standard out.
	printf "%s" "${return_status_prompt}"
}

export PS1='\fmt[\fD.$(_check_last_return_status)\fmt]\fd \fmt>\fd\fyt>\fd\fct>\fd '
```
In my mind this should work, but it doesn't. When I run a shell, `return_status` shows as blank. If I remove the colorization codes from `return_status_prompt`, then `return_status` will show properly. I have not been able to figure out why this is and wanted to ask here if this is some byproduct of an intentional feature, or if I found a bug. Thank you for your time.

This is wholey unrelated, but I also have set the `COMMAND_NOT_FOUND` handler to print an insult when a command is not found.
```
_print_insult() {
	## Print an insult.
	current_insult=$(sort -uR "${YASH_INSULTS_FILE}" | head -n 1)
	printf "%s" "${current_insult}"
}

COMMAND_NOT_FOUND_HANDLER=()
COMMAND_NOT_FOUND_HANDLER=_print_insult
```
It works, but I want to disable the standard `yash: no such command` output. Is there a way to do this?

**Comment by magicant** [comment]
Every font style notation starting with `\f` should end with a dot.

Instead of:

```
		return_status_prompt="\frt${return_status}\fd"
```

Try this:

```
		return_status_prompt="\frt.${return_status}\fd."
```

----

To suppress the `no such command` message, you can add `HANDLED=1` in the handler function. You might also want to add `return 127` to keep the non-zero exit status.

**Reply by FOSSilizedDaemon** [reply]
Aha! That makes sense. Quick question, just want to avoid unneeded discussion threads. How do you keep several `COMMAND_NOT_FOUND_HANDLER` functions from running? I have a few, `_walk()`, which will `cd ../` for every `..` entered, and `_print_insult()`, which insults me if command-not-found. I have _walk() first in the list, but if _walk() runs successfully I don't want `_print_insult()` to run.

I am working on porting my `zsh` configuration to POSIX shell and so far it has been pretty easy thanks your hard work :D

**Reply by magicant** [reply]
Assuming you have an array of handlers like `COMMAND_NOT_FOUND_HANDLER=(_walk _print_insult)`, you can quit the handler with `break -i`. See the `_autocd` function in the sample script: https://github.com/magicant/yash/blob/3f6f03967a34ccfc53a04b87862f9664b3e98851/share/initialization/common#L326-L332

**Reply by FOSSilizedDaemon** [reply]
Thank you :D
